src/types/complex.rs

```diff
@@ -19,118 +19,224 @@ pyobject_native_type!(
         ffi::PyComplex_FromCComplex(operation(l_val, r_val))
     }
 
-    impl<'py> Add for &'py PyComplex {
-        type Output = &'py PyComplex;
-        fn add(self, other: &'py PyComplex) -> &'py PyComplex {
-            unsafe {
-                self.py()
-                    .from_owned_ptr(complex_operation(self, other, ffi::_Py_c_sum))
+    macro_rules! bin_ops {
+        ($trait:ident, $fn:ident, $op:tt, $ffi:path) => {
+            impl<'py> $trait for Borrowed<'_, 'py, PyComplex> {
+                type Output = Bound<'py, PyComplex>;
+                fn $fn(self, other: Self) -> Self::Output {
+                    unsafe {
+                        complex_operation(self, other, $ffi)
+                            .assume_owned(self.py())
+                            .downcast_into_unchecked()
+                    }
+                }
             }
-        }
+
+            impl<'py> $trait for &'py PyComplex {
+                type Output = &'py PyComplex;
+                fn $fn(self, other: &'py PyComplex) -> &'py PyComplex {
+                    (self.as_borrowed() $op other.as_borrowed()).into_gil_ref()
+                }
+            }
+
+            impl<'py> $trait for &Bound<'py, PyComplex> {
+                type Output = Bound<'py, PyComplex>;
+                fn $fn(self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+                    self.as_borrowed() $op other.as_borrowed()
+                }
+            }
+
+            impl<'py> $trait<Bound<'py, PyComplex>> for &Bound<'py, PyComplex> {
+                type Output = Bound<'py, PyComplex>;
+                fn $fn(self, other: Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+                    self.as_borrowed() $op other.as_borrowed()
+                }
+            }
+
+            impl<'py> $trait for Bound<'py, PyComplex> {
+                type Output = Bound<'py, PyComplex>;
+                fn $fn(self, other: Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+                    self.as_borrowed() $op other.as_borrowed()
+                }
+            }
+
+            impl<'py> $trait<&Self> for Bound<'py, PyComplex> {
+                type Output = Bound<'py, PyComplex>;
+                fn $fn(self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+                    self.as_borrowed() $op other.as_borrowed()
+                }
+            }
+        };
     }
 
-    impl<'py> Sub for &'py PyComplex {
+    bin_ops!(Add, add, +, ffi::_Py_c_sum);
+    bin_ops!(Sub, sub, -, ffi::_Py_c_diff);
+    bin_ops!(Mul, mul, *, ffi::_Py_c_prod);
+    bin_ops!(Div, div, /, ffi::_Py_c_quot);
+
+    impl<'py> Neg for &'py PyComplex {
         type Output = &'py PyComplex;
-        fn sub(self, other: &'py PyComplex) -> &'py PyComplex {
+        fn neg(self) -> &'py PyComplex {
             unsafe {
+                let val = (*(self.as_ptr() as *mut ffi::PyComplexObject)).cval;
```

Comment:
While you're touching up this code, in new code I have a slight preference to use `.cast` instead of `as`:

```suggestion
                let val = (*(self.as_ptr().cast::<ffi::PyComplexObject>())).cval;
```

... it might even be possible to omit some parentheses:

```suggestion
                let val = (*self.as_ptr().cast::<ffi::PyComplexObject>()).cval;
```